sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala

```diff
@@ -1121,23 +1121,23 @@ class SQLContext(@transient val sparkContext: SparkContext)
     def assertAnalyzed(): Unit = analyzer.checkAnalysis(analyzed)
 
     lazy val analyzed: LogicalPlan = analyzer.execute(logical)
-    lazy val withCachedData: LogicalPlan = {
+    def withCachedData: LogicalPlan = {
       assertAnalyzed()
       cacheManager.useCachedData(analyzed)
     }
-    lazy val optimizedPlan: LogicalPlan = optimizer.execute(withCachedData)
+    def optimizedPlan: LogicalPlan = optimizer.execute(withCachedData)
 
     // TODO: Don't just pick the first one...
-    lazy val sparkPlan: SparkPlan = {
+    def sparkPlan: SparkPlan = {
       SparkPlan.currentContext.set(self)
       planner(optimizedPlan).next()
     }
     // executedPlan should not be used to initialize any SparkPlan. It should be
     // only used for execution.
-    lazy val executedPlan: SparkPlan = prepareForExecution.execute(sparkPlan)
+    def executedPlan: SparkPlan = prepareForExecution.execute(sparkPlan)
 
     /** Internal version of the RDD. Avoids copies and has no schema */
-    lazy val toRdd: RDD[Row] = executedPlan.execute()
+    def toRdd: RDD[Row] = executedPlan.execute()
```

Comment:
The snippet you just mentioned should be OK. Even in master code, the `CREATE TABLE` command will also be executed twice. But the following doesn't:

```scala
val df = sql("CREAT TABLE ...")
df.collect()
df.collect()
```

The command is executed while constructing the result DataFrame.